ldap/client/api/src/main/java/org/apache/directory/ldap/client/api/LdapNetworkConnection.java

```diff
@@ -552,125 +586,117 @@ public boolean connect() throws LdapException
             return true;
         }
 
-        // Create the connector if needed
-        if ( connector == null )
+        // We look up all the IP addresses behind the given DNS name, since one name might point to both A (IPv4) and AAAA (IPv6) records.
+        InetAddress[] addresses;
+        try
         {
-            createConnector();
+            addresses = InetAddress.getAllByName( config.getLdapHost() );
+        }
+        catch ( UnknownHostException e )
+        {
+            throw new LdapOtherException( e.getMessage(), e );
         }
 
-        // Build the connection address
-        SocketAddress address = new InetSocketAddress( config.getLdapHost(), config.getLdapPort() );
-
-        // And create the connection future
-        timeout = config.getTimeout();
-        long maxRetry = System.currentTimeMillis() + timeout;
+        // We try to initialize a socket connection with all the IPs simultaneously, picking the first that succeeds and dropping the others.
         ConnectFuture connectionFuture = null;
-
-        while ( maxRetry > System.currentTimeMillis() )
+        LdapNetworkConnectionFuture connFuture = null;
+        final long giveUpAfter = System.currentTimeMillis() + timeout * 2;
+        ExecutorService executor = Executors.newFixedThreadPool( addresses.length + 1 ); // Just in case of overlapping.
+        while ( connFuture == null && giveUpAfter > System.currentTimeMillis() )
```

Comment:
looks like a CPU heavy loop. I'd look for something like "wait until one of the Futures is done".
This could be helpful: http://stackoverflow.com/questions/19348248/waiting-on-a-list-of-future